assets/model_monitoring/components/src/action_analyzer/contracts/action_detectors/low_retreival_score_index_action_detector.py

```diff
@@ -3,49 +3,143 @@
+            2. extract extra fields from the root span for action.
+            3. convert the dataframe from trace level to span level.
+
         Args:
             df(pandas.DataFrame): input pandas dataframe.
 
         Returns:
             pandas.DataFrame: preprocessed pandas dataframe.
         """
-        pass
+        missed_metrics = get_missed_metrics(self.violated_metrics, df.columns.tolist())
+
+        if missed_metrics != []:
+            df = calculate_e2e_metrics(df, missed_metrics)
 
-    def detect(self, df: pandas.DataFrame, llm_client: LLMClient) -> list(Action):
+        return extract_fields_from_debugging_info(df, self.index_id)
+
+    def detect(self, df: pandas.DataFrame, llm_client: LLMClient, aml_deployment_id=None) -> list[Action]:
         """Detect the action.
 
         Args:
             df(pandas.DataFrame): input pandas dataframe.
             llm_client(LLMClient): LLM client used to get some llm scores/info for action.
+            aml_deployment_id(str): (Optional) aml deployment id for the action.
 
         Returns:
-            list(Action): list of actions.
+            list[Action]: list of actions.
         """
-        pass
+        # get llm retrieval score
+        df[INDEX_SCORE_LLM_COLUMN] = df.apply(get_retrieval_score, axis=1, args=(llm_client,))
+        df = df[df[INDEX_SCORE_LLM_COLUMN] != DEFAULT_LLM_SCORE]
+
+        action_list = []
+        for metric in self.violated_metrics:
+            low_retrieval_score_df = df[(df[metric] < METRICS_VIOLATION_THRESHOLD) &
+                                        (df[INDEX_SCORE_LLM_COLUMN] < METRICS_VIOLATION_THRESHOLD)]
+            high_retrieval_score_df = df[(df[metric] >= GOOD_METRICS_THRESHOLD) &
+                                         (df[INDEX_SCORE_LLM_COLUMN] >= GOOD_METRICS_THRESHOLD)]
+            # generate action only low retrieval score query ratio above the threshold
+            low_retrieval_score_query_ratio = len(low_retrieval_score_df)/len(df)
+            if low_retrieval_score_query_ratio >= LOW_RETRIEVAL_SCORE_QUERY_RATIO_THRESHOLD:
+                print(f"Generating action for metric {metric}. \
+                The low retrieval score query ratio is {low_retrieval_score_query_ratio}.")
+                # use the low retrieval score query ratio as confidence
+                action = self.generate_action(llm_client,
+                                              metric,
+                                              low_retrieval_score_query_ratio,
+                                              low_retrieval_score_df,
+                                              high_retrieval_score_df,
+                                              aml_deployment_id)
+                print(f"Positive sample size: {len(action.positive_samples)}.")
+                print(f"Negative sample size: {len(action.negative_samples)}.")
+                action_list.append(action)
```

Comment:
for different metrics, would we have very similar actions? do we also have de-dup logic here

Reply:
This will be returned to the main function, and we will do the dedup altogether there. The idea is: for a same index, 
we might have duplicated actions for different metrics and different detectors. We will do the dedup at the same place.